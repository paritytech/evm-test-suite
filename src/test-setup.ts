import { killProcessOnPort, waitForHealth } from './util.ts'

export type Env = 'geth' | 'eth-rpc'

let processes: Deno.ChildProcess[] = []
let envs: Env[] = []

export async function setupTests(): Promise<Env[]> {
    let useGeth = !!Deno.env.get('USE_GETH')
    let useEthRpc = !!Deno.env.get('USE_ETH_RPC')

    if (Deno.env.get('START_GETH')) {
        useGeth = true
        const geth = Deno.env.get('GETH_BIN') ?? 'geth'
        const gethArgs = [
            '--http',
            '--http.api',
            'web3,eth,debug,personal,net',
            '--http.port',
            '8546',
            '--dev',
            '--verbosity',
            '0',
        ]

        await killProcessOnPort(8546)
        console.log('ðŸš€ Start geth...')
        const gethProcess = new Deno.Command(geth, {
            args: gethArgs,
            stdout: 'null',
            stderr: 'null',
        }).spawn()
        processes.push(gethProcess)
        await waitForHealth('http://localhost:8546').catch(() => {})
    }

    if (Deno.env.get('START_REVIVE_DEV_NODE')) {
        const devNodeArgs = [
            '--dev',
            '-l=error,evm=debug,sc_rpc_server=info,runtime::revive=debug',
        ]

        await killProcessOnPort(9944)
        const nodePath = Deno.env.get('REVIVE_DEV_NODE_PATH') ??
            `${Deno.env.get('HOME')}/polkadot-sdk/target/debug/revive-dev-node`
        console.log('ðŸš€ Start dev-node ...')
        const devNodeProcess = new Deno.Command(nodePath, {
            args: devNodeArgs,
            stdout: 'null',
            stderr: 'null',
        }).spawn()
        processes.push(devNodeProcess)
    }

    if (Deno.env.get('START_ETH_RPC')) {
        useEthRpc = true

        // Run eth-rpc on 8545
        const ethRpcArgs = [
            '--dev',
            '--node-rpc-url=ws://localhost:9944',
            '-l=rpc-metrics=debug,eth-rpc=debug',
        ]

        await killProcessOnPort(8545)
        const adapterPath = Deno.env.get('ETH_RPC_PATH') ??
            `${Deno.env.get('HOME')}/polkadot-sdk/target/debug/eth-rpc`
        console.log('ðŸš€ Start eth-rpc ...')
        const ethRpcProcess = new Deno.Command(adapterPath, {
            args: ethRpcArgs,
            stdout: 'null',
            stderr: 'null',
        }).spawn()
        processes.push(ethRpcProcess)
        await waitForHealth('http://localhost:8545').catch(() => {})
    }

    envs = [
        ...(useGeth ? ['geth' as const] : []),
        ...(useEthRpc ? ['eth-rpc' as const] : []),
    ]

    if (envs.length === 0) {
        throw new Error('No environment started. Set USE_GETH or USE_ETH_RPC')
    }

    return envs
}

export function cleanupTests() {
    if (processes.length === 0) {
        return
    }
    console.log('ðŸ”Œ Shutting down servers...')
    for (const proc of processes) {
        try {
            proc.kill('SIGTERM')
        } catch {
            // Process might already be dead
        }
    }
    processes = []
}

export function getEnvs(): Env[] {
    return envs
}
