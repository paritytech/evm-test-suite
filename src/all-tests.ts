// Global test setup - runs once before all tests
import { cleanupTests, setupTests } from './test-setup.ts'

await setupTests()
await import('./methods.test.ts')
await import('./errors.test.ts')
await import('./gas.test.ts')
await import('./others.test.ts')
await import('./tracing-call-trace.test.ts')
await import('./tracing-prestate.test.ts')

globalThis.addEventListener('unload', () => {
    cleanupTests()
})
