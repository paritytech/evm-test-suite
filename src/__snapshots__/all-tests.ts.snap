export const snapshot = {};

snapshot[`debug_traceTransaction 1`] = `
{
  calls: [
    {
      from: "<contract_addr>",
      gas: "0x42",
      gasUsed: "0x42",
      input: "0x",
      to: "<contract_callee_addr>",
      type: "CALL",
      value: "0x2386f26fc10000",
    },
    {
      from: "<contract_addr>",
      gas: "0x42",
      gasUsed: "0x42",
      input: "0xa329e8de0000000000000000000000000000000000000000000000000000000000000002",
      logs: [
        {
          address: "<contract_callee_addr>",
          data: "0x0000000000000000000000000000000000000000000000000000000000000002",
          position: "0x0",
          topics: [
            "0xa07465e8ec189714f79f3786a8f616baf78ebd9cb1769bd61f18de45f2567360",
          ],
        },
      ],
      to: "<contract_callee_addr>",
      type: "CALL",
      value: "0x0",
    },
    {
      error: "execution reverted",
      from: "<contract_addr>",
      gas: "0x42",
      gasUsed: "0x42",
      input: "0xd1b96663",
      output: "0x08c379a00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000001a546869732066756e6374696f6e20616c77617973206661696c73000000000000",
      revertReason: "This function always fails",
      to: "<contract_callee_addr>",
      type: "CALL",
      value: "0x2386f26fc10000",
    },
    {
      calls: [
        {
          from: "<contract_addr>",
          gas: "0x42",
          gasUsed: "0x42",
          input: "0x",
          to: "<contract_callee_addr>",
          type: "CALL",
          value: "0x2386f26fc10000",
        },
        {
          from: "<contract_addr>",
          gas: "0x42",
          gasUsed: "0x42",
          input: "0xa329e8de0000000000000000000000000000000000000000000000000000000000000001",
          logs: [
            {
              address: "<contract_callee_addr>",
              data: "0x0000000000000000000000000000000000000000000000000000000000000001",
              position: "0x0",
              topics: [
                "0xa07465e8ec189714f79f3786a8f616baf78ebd9cb1769bd61f18de45f2567360",
              ],
            },
          ],
          to: "<contract_callee_addr>",
          type: "CALL",
          value: "0x0",
        },
        {
          error: "execution reverted",
          from: "<contract_addr>",
          gas: "0x42",
          gasUsed: "0x42",
          input: "0xd1b96663",
          output: "0x08c379a00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000001a546869732066756e6374696f6e20616c77617973206661696c73000000000000",
          revertReason: "This function always fails",
          to: "<contract_callee_addr>",
          type: "CALL",
          value: "0x2386f26fc10000",
        },
        {
          from: "<contract_addr>",
          gas: "0x42",
          gasUsed: "0x42",
          input: "0x95805dad0000000000000000000000000000000000000000000000000000000000000000",
          to: "<contract_addr>",
          type: "CALL",
          value: "0x0",
        },
      ],
      from: "<contract_addr>",
      gas: "0x42",
      gasUsed: "0x42",
      input: "0x95805dad0000000000000000000000000000000000000000000000000000000000000001",
      logs: [
        {
          address: "<contract_addr>",
          data: "0x0000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000066265666f72650000000000000000000000000000000000000000000000000000",
          position: "0x1",
          topics: [
            "0x25d760f35a7a9cb2bffd2ea8756913655b3786c642f300a702e2934062763549",
          ],
        },
        {
          address: "<contract_addr>",
          data: "0x0000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000056166746572000000000000000000000000000000000000000000000000000000",
          position: "0x2",
          topics: [
            "0x25d760f35a7a9cb2bffd2ea8756913655b3786c642f300a702e2934062763549",
          ],
        },
      ],
      to: "<contract_addr>",
      type: "CALL",
      value: "0x0",
    },
  ],
  from: "<caller>",
  gas: "0x42",
  gasUsed: "0x42",
  input: "0x95805dad0000000000000000000000000000000000000000000000000000000000000002",
  logs: [
    {
      address: "<contract_addr>",
      data: "0x0000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000066265666f72650000000000000000000000000000000000000000000000000000",
      position: "0x1",
      topics: [
        "0x25d760f35a7a9cb2bffd2ea8756913655b3786c642f300a702e2934062763549",
      ],
    },
    {
      address: "<contract_addr>",
      data: "0x0000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000056166746572000000000000000000000000000000000000000000000000000000",
      position: "0x2",
      topics: [
        "0x25d760f35a7a9cb2bffd2ea8756913655b3786c642f300a702e2934062763549",
      ],
    },
  ],
  to: "<contract_addr>",
  type: "CALL",
  value: "0x0",
}
`;

snapshot[`debug_deploy_traceTransaction 1`] = `
{
  from: "<caller>",
  gas: "0x42",
  gasUsed: "0x42",
  input: "<code>",
  to: "<contract_callee_addr>",
  type: "CREATE",
  value: "0x0",
}
`;

snapshot[`debug_traceBlock 1`] = `
[
  {
    result: {
      calls: [
        {
          from: "<contract_addr>",
          gas: "0x42",
          gasUsed: "0x42",
          input: "0x",
          to: "<contract_callee_addr>",
          type: "CALL",
          value: "0x2386f26fc10000",
        },
        {
          from: "<contract_addr>",
          gas: "0x42",
          gasUsed: "0x42",
          input: "0xa329e8de0000000000000000000000000000000000000000000000000000000000000002",
          logs: [
            {
              address: "<contract_callee_addr>",
              data: "0x0000000000000000000000000000000000000000000000000000000000000002",
              position: "0x0",
              topics: [
                "0xa07465e8ec189714f79f3786a8f616baf78ebd9cb1769bd61f18de45f2567360",
              ],
            },
          ],
          to: "<contract_callee_addr>",
          type: "CALL",
          value: "0x0",
        },
        {
          error: "execution reverted",
          from: "<contract_addr>",
          gas: "0x42",
          gasUsed: "0x42",
          input: "0xd1b96663",
          output: "0x08c379a00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000001a546869732066756e6374696f6e20616c77617973206661696c73000000000000",
          revertReason: "This function always fails",
          to: "<contract_callee_addr>",
          type: "CALL",
          value: "0x2386f26fc10000",
        },
        {
          calls: [
            {
              from: "<contract_addr>",
              gas: "0x42",
              gasUsed: "0x42",
              input: "0x",
              to: "<contract_callee_addr>",
              type: "CALL",
              value: "0x2386f26fc10000",
            },
            {
              from: "<contract_addr>",
              gas: "0x42",
              gasUsed: "0x42",
              input: "0xa329e8de0000000000000000000000000000000000000000000000000000000000000001",
              logs: [
                {
                  address: "<contract_callee_addr>",
                  data: "0x0000000000000000000000000000000000000000000000000000000000000001",
                  position: "0x0",
                  topics: [
                    "0xa07465e8ec189714f79f3786a8f616baf78ebd9cb1769bd61f18de45f2567360",
                  ],
                },
              ],
              to: "<contract_callee_addr>",
              type: "CALL",
              value: "0x0",
            },
            {
              error: "execution reverted",
              from: "<contract_addr>",
              gas: "0x42",
              gasUsed: "0x42",
              input: "0xd1b96663",
              output: "0x08c379a00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000001a546869732066756e6374696f6e20616c77617973206661696c73000000000000",
              revertReason: "This function always fails",
              to: "<contract_callee_addr>",
              type: "CALL",
              value: "0x2386f26fc10000",
            },
            {
              from: "<contract_addr>",
              gas: "0x42",
              gasUsed: "0x42",
              input: "0x95805dad0000000000000000000000000000000000000000000000000000000000000000",
              to: "<contract_addr>",
              type: "CALL",
              value: "0x0",
            },
          ],
          from: "<contract_addr>",
          gas: "0x42",
          gasUsed: "0x42",
          input: "0x95805dad0000000000000000000000000000000000000000000000000000000000000001",
          logs: [
            {
              address: "<contract_addr>",
              data: "0x0000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000066265666f72650000000000000000000000000000000000000000000000000000",
              position: "0x1",
              topics: [
                "0x25d760f35a7a9cb2bffd2ea8756913655b3786c642f300a702e2934062763549",
              ],
            },
            {
              address: "<contract_addr>",
              data: "0x0000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000056166746572000000000000000000000000000000000000000000000000000000",
              position: "0x2",
              topics: [
                "0x25d760f35a7a9cb2bffd2ea8756913655b3786c642f300a702e2934062763549",
              ],
            },
          ],
          to: "<contract_addr>",
          type: "CALL",
          value: "0x0",
        },
      ],
      from: "<caller>",
      gas: "0x42",
      gasUsed: "0x42",
      input: "0x95805dad0000000000000000000000000000000000000000000000000000000000000002",
      logs: [
        {
          address: "<contract_addr>",
          data: "0x0000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000066265666f72650000000000000000000000000000000000000000000000000000",
          position: "0x1",
          topics: [
            "0x25d760f35a7a9cb2bffd2ea8756913655b3786c642f300a702e2934062763549",
          ],
        },
        {
          address: "<contract_addr>",
          data: "0x0000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000056166746572000000000000000000000000000000000000000000000000000000",
          position: "0x2",
          topics: [
            "0x25d760f35a7a9cb2bffd2ea8756913655b3786c642f300a702e2934062763549",
          ],
        },
      ],
      to: "<contract_addr>",
      type: "CALL",
      value: "0x0",
    },
    txHash: "<hash>",
  },
]
`;

snapshot[`debug_traceCall 1`] = `
{
  calls: [
    {
      from: "<contract_addr>",
      gas: "0x42",
      gasUsed: "0x42",
      input: "0x",
      to: "<contract_callee_addr>",
      type: "CALL",
      value: "0x2386f26fc10000",
    },
    {
      from: "<contract_addr>",
      gas: "0x42",
      gasUsed: "0x42",
      input: "0xa329e8de0000000000000000000000000000000000000000000000000000000000000002",
      logs: [
        {
          address: "<contract_callee_addr>",
          data: "0x0000000000000000000000000000000000000000000000000000000000000002",
          position: "0x0",
          topics: [
            "0xa07465e8ec189714f79f3786a8f616baf78ebd9cb1769bd61f18de45f2567360",
          ],
        },
      ],
      to: "<contract_callee_addr>",
      type: "CALL",
      value: "0x0",
    },
    {
      error: "execution reverted",
      from: "<contract_addr>",
      gas: "0x42",
      gasUsed: "0x42",
      input: "0xd1b96663",
      output: "0x08c379a00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000001a546869732066756e6374696f6e20616c77617973206661696c73000000000000",
      revertReason: "This function always fails",
      to: "<contract_callee_addr>",
      type: "CALL",
      value: "0x2386f26fc10000",
    },
    {
      calls: [
        {
          from: "<contract_addr>",
          gas: "0x42",
          gasUsed: "0x42",
          input: "0x",
          to: "<contract_callee_addr>",
          type: "CALL",
          value: "0x2386f26fc10000",
        },
        {
          from: "<contract_addr>",
          gas: "0x42",
          gasUsed: "0x42",
          input: "0xa329e8de0000000000000000000000000000000000000000000000000000000000000001",
          logs: [
            {
              address: "<contract_callee_addr>",
              data: "0x0000000000000000000000000000000000000000000000000000000000000001",
              position: "0x0",
              topics: [
                "0xa07465e8ec189714f79f3786a8f616baf78ebd9cb1769bd61f18de45f2567360",
              ],
            },
          ],
          to: "<contract_callee_addr>",
          type: "CALL",
          value: "0x0",
        },
        {
          error: "execution reverted",
          from: "<contract_addr>",
          gas: "0x42",
          gasUsed: "0x42",
          input: "0xd1b96663",
          output: "0x08c379a00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000001a546869732066756e6374696f6e20616c77617973206661696c73000000000000",
          revertReason: "This function always fails",
          to: "<contract_callee_addr>",
          type: "CALL",
          value: "0x2386f26fc10000",
        },
        {
          from: "<contract_addr>",
          gas: "0x42",
          gasUsed: "0x42",
          input: "0x95805dad0000000000000000000000000000000000000000000000000000000000000000",
          to: "<contract_addr>",
          type: "CALL",
          value: "0x0",
        },
      ],
      from: "<contract_addr>",
      gas: "0x42",
      gasUsed: "0x42",
      input: "0x95805dad0000000000000000000000000000000000000000000000000000000000000001",
      logs: [
        {
          address: "<contract_addr>",
          data: "0x0000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000066265666f72650000000000000000000000000000000000000000000000000000",
          position: "0x1",
          topics: [
            "0x25d760f35a7a9cb2bffd2ea8756913655b3786c642f300a702e2934062763549",
          ],
        },
        {
          address: "<contract_addr>",
          data: "0x0000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000056166746572000000000000000000000000000000000000000000000000000000",
          position: "0x2",
          topics: [
            "0x25d760f35a7a9cb2bffd2ea8756913655b3786c642f300a702e2934062763549",
          ],
        },
      ],
      to: "<contract_addr>",
      type: "CALL",
      value: "0x0",
    },
  ],
  from: "0x0000000000000000000000000000000000000000",
  gas: "0x42",
  gasUsed: "0x42",
  input: "0x95805dad0000000000000000000000000000000000000000000000000000000000000002",
  logs: [
    {
      address: "<contract_addr>",
      data: "0x0000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000066265666f72650000000000000000000000000000000000000000000000000000",
      position: "0x1",
      topics: [
        "0x25d760f35a7a9cb2bffd2ea8756913655b3786c642f300a702e2934062763549",
      ],
    },
    {
      address: "<contract_addr>",
      data: "0x0000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000056166746572000000000000000000000000000000000000000000000000000000",
      position: "0x2",
      topics: [
        "0x25d760f35a7a9cb2bffd2ea8756913655b3786c642f300a702e2934062763549",
      ],
    },
  ],
  to: "<contract_addr>",
  type: "CALL",
  value: "0x0",
}
`;

snapshot[`write_storage.diff 1`] = `
{
  post: {
    "<caller_addr>": {
      nonce: "<nonce>",
    },
    "<contract_addr>": {
      storage: {
        "0x0000000000000000000000000000000000000000000000000000000000000000": "0x00000000000000000000000000000000000000000000000000000000000007e9",
        "0x0000000000000000000000000000000000000000000000000000000000000002": "0x77726974655f73746f726167650000000000000000000000000000000000001a",
      },
    },
  },
  pre: {
    "<caller_addr>": {
      balance: "<balance>",
      nonce: "<nonce>",
    },
    "<contract_addr>": {
      balance: "<balance>",
      code: "<code>",
      nonce: "<nonce>",
      storage: {
        "0x0000000000000000000000000000000000000000000000000000000000000000": "0x000000000000000000000000000000000000000000000000000000000000002a",
        "0x0000000000000000000000000000000000000000000000000000000000000002": "0x696e697400000000000000000000000000000000000000000000000000000008",
      },
    },
  },
}
`;

snapshot[`write_storage.no_diff 1`] = `
{
  "<caller_addr>": {
    balance: "<balance>",
    nonce: "<nonce>",
  },
  "<contract_addr>": {
    balance: "<balance>",
    code: "<code>",
    nonce: "<nonce>",
    storage: {
      "0x0000000000000000000000000000000000000000000000000000000000000000": "0x000000000000000000000000000000000000000000000000000000000000002a",
      "0x0000000000000000000000000000000000000000000000000000000000000002": "0x696e697400000000000000000000000000000000000000000000000000000008",
    },
  },
}
`;

snapshot[`write_storage_from_0.diff 1`] = `
{
  post: {
    "<contract_addr>": {
      storage: {
        "0x0000000000000000000000000000000000000000000000000000000000000000": "0x0000000000000000000000000000000000000000000000000000000000000002",
        "0x0000000000000000000000000000000000000000000000000000000000000002": "0x77726974655f73746f726167655f66726f6d5f30000000000000000000000028",
      },
    },
  },
  pre: {
    "<contract_addr>": {
      balance: "<balance>",
      code: "<code>",
      nonce: "<nonce>",
      storage: {
        "0x0000000000000000000000000000000000000000000000000000000000000000": "0x000000000000000000000000000000000000000000000000000000000000002a",
        "0x0000000000000000000000000000000000000000000000000000000000000002": "0x696e697400000000000000000000000000000000000000000000000000000008",
      },
    },
  },
}
`;

snapshot[`write_storage_from_0.no_diff 1`] = `
{
  "<contract_addr>": {
    balance: "<balance>",
    code: "<code>",
    nonce: "<nonce>",
    storage: {
      "0x0000000000000000000000000000000000000000000000000000000000000000": "0x000000000000000000000000000000000000000000000000000000000000002a",
      "0x0000000000000000000000000000000000000000000000000000000000000002": "0x696e697400000000000000000000000000000000000000000000000000000008",
    },
  },
}
`;

snapshot[`read_storage.diff 1`] = `
{
  post: {
    "<caller_addr>": {
      nonce: "<nonce>",
    },
  },
  pre: {
    "<caller_addr>": {
      balance: "<balance>",
      nonce: "<nonce>",
    },
  },
}
`;

snapshot[`read_storage.no_diff 1`] = `
{
  "<caller_addr>": {
    balance: "<balance>",
    nonce: "<nonce>",
  },
  "<contract_addr>": {
    balance: "<balance>",
    code: "<code>",
    nonce: "<nonce>",
    storage: {
      "0x0000000000000000000000000000000000000000000000000000000000000000": "0x000000000000000000000000000000000000000000000000000000000000002a",
    },
  },
}
`;

snapshot[`deposit.diff 1`] = `
{
  post: {
    "<caller_addr>": {
      balance: "<balance>",
      nonce: "<nonce>",
    },
    "<contract_addr>": {
      balance: "<balance>",
      storage: {
        "<wallet_balance>": "0x00000000000000000000000000000000000000000000000098a7d9b8314c0000",
      },
    },
  },
  pre: {
    "<caller_addr>": {
      balance: "<balance>",
      nonce: "<nonce>",
    },
    "<contract_addr>": {
      balance: "<balance>",
      code: "<code>",
      nonce: "<nonce>",
      storage: {
        "<wallet_balance>": "0x0000000000000000000000000000000000000000000000008ac7230489e80000",
      },
    },
  },
}
`;

snapshot[`deposit.no_diff 1`] = `
{
  "<caller_addr>": {
    balance: "<balance>",
    nonce: "<nonce>",
  },
  "<contract_addr>": {
    balance: "<balance>",
    code: "<code>",
    nonce: "<nonce>",
    storage: {
      "<wallet_balance>": "0x0000000000000000000000000000000000000000000000008ac7230489e80000",
    },
  },
}
`;

snapshot[`withdraw.diff 1`] = `
{
  post: {
    "<caller_addr>": {
      balance: "<balance>",
      nonce: "<nonce>",
    },
    "<contract_addr>": {
      balance: "<balance>",
      storage: {
        "<wallet_balance>": "0x00000000000000000000000000000000000000000000000098a7d9b8314c0000",
      },
    },
  },
  pre: {
    "<caller_addr>": {
      balance: "<balance>",
      nonce: "<nonce>",
    },
    "<contract_addr>": {
      balance: "<balance>",
      code: "<code>",
      nonce: "<nonce>",
      storage: {
        "<wallet_balance>": "0x0000000000000000000000000000000000000000000000008ac7230489e80000",
      },
    },
  },
}
`;

snapshot[`withdraw.no_diff 1`] = `
{
  "<caller_addr>": {
    balance: "<balance>",
    nonce: "<nonce>",
  },
  "<contract_addr>": {
    balance: "<balance>",
    code: "<code>",
    nonce: "<nonce>",
    storage: {
      "<wallet_balance>": "0x0000000000000000000000000000000000000000000000008ac7230489e80000",
    },
  },
}
`;

snapshot[`get_balance.diff 1`] = `
{
  post: {
    "<caller_addr>": {
      nonce: "<nonce>",
    },
  },
  pre: {
    "<caller_addr>": {
      balance: "<balance>",
      nonce: "<nonce>",
    },
  },
}
`;

snapshot[`get_balance.no_diff 1`] = `
{
  "<caller_addr>": {
    balance: "<balance>",
    nonce: "<nonce>",
  },
  "<contract_addr>": {
    balance: "<balance>",
    code: "<code>",
    nonce: "<nonce>",
  },
}
`;

snapshot[`get_external_balance.diff 1`] = `
{
  post: {
    "<caller_addr>": {
      nonce: "<nonce>",
    },
  },
  pre: {
    "<caller_addr>": {
      balance: "<balance>",
      nonce: "<nonce>",
    },
  },
}
`;

snapshot[`get_external_balance.no_diff 1`] = `
{
  "<caller_addr>": {
    balance: "<balance>",
    nonce: "<nonce>",
  },
  "<contract_addr>": {
    balance: "<balance>",
    code: "<code>",
    nonce: "<nonce>",
  },
}
`;

snapshot[`instantiate_child.diff 1`] = `
{
  post: {
    "<caller_addr>": {
      nonce: "<nonce>",
    },
  },
  pre: {
    "<caller_addr>": {
      balance: "<balance>",
      nonce: "<nonce>",
    },
  },
}
`;

snapshot[`instantiate_child.no_diff 1`] = `
{
  "<caller_addr>": {
    balance: "<balance>",
    nonce: "<nonce>",
  },
  "<contract_addr>": {
    balance: "<balance>",
    code: "<code>",
    nonce: "<nonce>",
  },
}
`;

snapshot[`call_contract.diff 1`] = `
{
  post: {
    "<caller_addr>": {
      nonce: "<nonce>",
    },
    "<contract_addr_2>": {
      storage: {
        "0x0000000000000000000000000000000000000000000000000000000000000000": "0x0000000000000000000000000000000000000000000000000000000000000002",
      },
    },
  },
  pre: {
    "<caller_addr>": {
      balance: "<balance>",
      nonce: "<nonce>",
    },
    "<contract_addr_2>": {
      balance: "<balance>",
      code: "<code>",
      nonce: "<nonce>",
      storage: {
        "0x0000000000000000000000000000000000000000000000000000000000000000": "0x0000000000000000000000000000000000000000000000000000000000000001",
      },
    },
  },
}
`;

snapshot[`call_contract.no_diff 1`] = `
{
  "<caller_addr>": {
    balance: "<balance>",
    nonce: "<nonce>",
  },
  "<contract_addr>": {
    balance: "<balance>",
    code: "<code>",
    nonce: "<nonce>",
  },
  "<contract_addr_2>": {
    balance: "<balance>",
    code: "<code>",
    nonce: "<nonce>",
    storage: {
      "0x0000000000000000000000000000000000000000000000000000000000000000": "0x0000000000000000000000000000000000000000000000000000000000000001",
    },
  },
}
`;

snapshot[`delegate_call_contract.diff 1`] = `
{
  post: {
    "<caller_addr>": {
      nonce: "<nonce>",
    },
  },
  pre: {
    "<caller_addr>": {
      balance: "<balance>",
      nonce: "<nonce>",
    },
  },
}
`;

snapshot[`delegate_call_contract.no_diff 1`] = `
{
  "<caller_addr>": {
    balance: "<balance>",
    nonce: "<nonce>",
  },
  "<contract_addr>": {
    balance: "<balance>",
    code: "<code>",
    nonce: "<nonce>",
  },
}
`;

snapshot[`trace_block_write_storage_twice.diff 1`] = `
[
  {
    result: {
      post: {
        "<caller_addr>": {
          balance: "<balance>",
          nonce: "<nonce>",
        },
        "<contract_addr>": {
          storage: {
            "0x0000000000000000000000000000000000000000000000000000000000000000": "0x000000000000000000000000000000000000000000000000000000000000002c",
            "0x0000000000000000000000000000000000000000000000000000000000000002": "0x77726974655f73746f726167655f747769636500000000000000000000000026",
          },
        },
      },
      pre: {
        "<caller_addr>": {
          balance: "<balance>",
          nonce: "<nonce>",
        },
        "<contract_addr>": {
          balance: "<balance>",
          code: "<code>",
          nonce: "<nonce>",
          storage: {
            "0x0000000000000000000000000000000000000000000000000000000000000000": "0x000000000000000000000000000000000000000000000000000000000000002a",
            "0x0000000000000000000000000000000000000000000000000000000000000002": "0x696e697400000000000000000000000000000000000000000000000000000008",
          },
        },
      },
    },
    txHash: "<tx_hash>",
  },
  {
    result: {
      post: {
        "<caller_addr>": {
          balance: "<balance>",
          nonce: "<nonce>",
        },
        "<contract_addr>": {
          storage: {
            "0x0000000000000000000000000000000000000000000000000000000000000000": "0x000000000000000000000000000000000000000000000000000000000000002b",
          },
        },
      },
      pre: {
        "<caller_addr>": {
          balance: "<balance>",
          nonce: "<nonce>",
        },
        "<contract_addr>": {
          balance: "<balance>",
          code: "<code>",
          nonce: "<nonce>",
          storage: {
            "0x0000000000000000000000000000000000000000000000000000000000000000": "0x000000000000000000000000000000000000000000000000000000000000002c",
          },
        },
      },
    },
    txHash: "<tx_hash>",
  },
]
`;

snapshot[`trace_tx_write_storage_twice.diff 1`] = `
{
  post: {
    "<caller_addr>": {
      balance: "<balance>",
      nonce: "<nonce>",
    },
    "<contract_addr>": {
      storage: {
        "0x0000000000000000000000000000000000000000000000000000000000000000": "0x000000000000000000000000000000000000000000000000000000000000002c",
        "0x0000000000000000000000000000000000000000000000000000000000000002": "0x77726974655f73746f726167655f747769636500000000000000000000000026",
      },
    },
  },
  pre: {
    "<caller_addr>": {
      balance: "<balance>",
      nonce: "<nonce>",
    },
    "<contract_addr>": {
      balance: "<balance>",
      code: "<code>",
      nonce: "<nonce>",
      storage: {
        "0x0000000000000000000000000000000000000000000000000000000000000000": "0x000000000000000000000000000000000000000000000000000000000000002a",
        "0x0000000000000000000000000000000000000000000000000000000000000002": "0x696e697400000000000000000000000000000000000000000000000000000008",
      },
    },
  },
}
`;

snapshot[`trace_block_write_storage_twice.no_diff 1`] = `
[
  {
    result: {
      "<caller_addr>": {
        balance: "<balance>",
        nonce: "<nonce>",
      },
      "<contract_addr>": {
        balance: "<balance>",
        code: "<code>",
        nonce: "<nonce>",
        storage: {
          "0x0000000000000000000000000000000000000000000000000000000000000000": "0x000000000000000000000000000000000000000000000000000000000000002b",
          "0x0000000000000000000000000000000000000000000000000000000000000002": "0x77726974655f73746f726167655f747769636500000000000000000000000026",
        },
      },
    },
    txHash: "<tx_hash>",
  },
  {
    result: {
      "<caller_addr>": {
        balance: "<balance>",
        nonce: "<nonce>",
      },
      "<contract_addr>": {
        balance: "<balance>",
        code: "<code>",
        nonce: "<nonce>",
        storage: {
          "0x0000000000000000000000000000000000000000000000000000000000000000": "0x000000000000000000000000000000000000000000000000000000000000002d",
          "0x0000000000000000000000000000000000000000000000000000000000000002": "0x77726974655f73746f726167655f747769636500000000000000000000000026",
        },
      },
    },
    txHash: "<tx_hash>",
  },
]
`;

snapshot[`trace_tx_write_storage_twice.no_diff 1`] = `
{
  "<caller_addr>": {
    balance: "<balance>",
    nonce: "<nonce>",
  },
  "<contract_addr>": {
    balance: "<balance>",
    code: "<code>",
    nonce: "<nonce>",
    storage: {
      "0x0000000000000000000000000000000000000000000000000000000000000000": "0x000000000000000000000000000000000000000000000000000000000000002b",
      "0x0000000000000000000000000000000000000000000000000000000000000002": "0x77726974655f73746f726167655f747769636500000000000000000000000026",
    },
  },
}
`;
